source/extensions/quic_listeners/quiche/envoy_quic_server_stream.cc

```diff
@@ -14,50 +14,118 @@
 #include "quiche/quic/core/http/quic_header_list.h"
 #include "quiche/quic/core/quic_session.h"
 #include "quiche/spdy/core/spdy_header_block.h"
-
+#include "extensions/quic_listeners/quiche/platform/quic_mem_slice_span_impl.h"
 #pragma GCC diagnostic pop
 
 #include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_server_session.h"
+
 #include "common/buffer/buffer_impl.h"
 #include "common/http/header_map_impl.h"
 #include "common/common/assert.h"
 
 namespace Envoy {
 namespace Quic {
 
+EnvoyQuicServerStream::EnvoyQuicServerStream(quic::QuicStreamId id, quic::QuicSpdySession* session,
+                                             quic::StreamType type)
+    : quic::QuicSpdyServerStreamBase(id, session, type),
+      EnvoyQuicStream(
+          2 * GetQuicFlag(FLAGS_quic_buffered_data_threshold),
+          [this]() { runLowWatermarkCallbacks(); }, [this]() { runHighWatermarkCallbacks(); }) {}
+
+EnvoyQuicServerStream::EnvoyQuicServerStream(quic::PendingStream* pending,
+                                             quic::QuicSpdySession* session, quic::StreamType type)
+    : quic::QuicSpdyServerStreamBase(pending, session, type),
+      EnvoyQuicStream(
+          session->config()->ReceivedInitialStreamFlowControlWindowBytes(),
+          [this]() { runLowWatermarkCallbacks(); }, [this]() { runHighWatermarkCallbacks(); }) {}
+
 void EnvoyQuicServerStream::encode100ContinueHeaders(const Http::HeaderMap& headers) {
   ASSERT(headers.Status()->value() == "100");
   encodeHeaders(headers, false);
 }
-void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& /*headers*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeHeaders(const Http::HeaderMap& headers, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeHeaders (end_stream={}) {}.", *this, end_stream, headers);
+  WriteHeaders(envoyHeadersToSpdyHeaderBlock(headers), end_stream, nullptr);
+  local_end_stream_ = end_stream;
 }
-void EnvoyQuicServerStream::encodeData(Buffer::Instance& /*data*/, bool /*end_stream*/) {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+
+void EnvoyQuicServerStream::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_STREAM_LOG(debug, "encodeData (end_stream={}) of {} bytes.", *this, end_stream,
+                   data.length());
+  local_end_stream_ = end_stream;
+  // This is counting not serialized bytes in the send buffer.
+  uint64_t bytes_to_send_old = BufferedDataBytes();
+  // QUIC stream must take all.
```

Comment:
Are we setting buffered_data_threshold_ high enough we think this is guaranteed?

Reply:
So the only relevant thing I find here is 
EnvoyQuicStream(2 * GetQuicFlag(FLAGS_quic_buffered_data_threshold) ...which is
EnvoyQuicStream(2 * 8 * 1024
which doesn't look remotely high enough 
what am I missing?  maybe the docs can call out where it's set since it's not (to me) obvious?